src/api-v1.js

```diff
@@ -580,6 +580,197 @@ api.declare({
+}, async function (req, res) {
+  let input = req.body;
+  let id = req.params.id;
+
+  input.lastModified = new Date();
+
+  // Authenticate request with parameterized scope
+  if (!req.satisfies({id})) {
+    return;
+  }
+
+  // Create amiSet
+  let amiSet;
+  try {
+    amiSet = await this.AmiSet.create({
+      id: id,
+      amis: input.amis,
+      lastModified: new Date(),
+    });
+  } catch (err) {
+    // We only catch EntityAlreadyExists errors
+    if (!err || err.code !== 'EntityAlreadyExists') {
+      throw err;
+    }
+    amiSet = await this.AmiSet.load({id});
+
+    // Check if it matches the existing amiSet
+    let match = [
+      'id',
+      'amis',
+    ].every((key) => {
+      return _.isEqual(amiSet[key], input[key]);
+    });
+
+    // If we don't have a match we return 409, otherwise we continue as this is
+    // is an idempotent operation.
+    if (!match) {
+      res.status(409).json({
+        error: 'AMI Set already exists with different definition',
+      });
+      return;
+    }
+  }
+  res.reply({outcome: 'success'});
+  return;
+
+});
+
+api.declare({
+  method: 'get',
+  route: '/ami-set/:id',
+  name: 'getAmiSet',
+  output: 'get-ami-set-response.json#',
+  deferAuth: true,
+  title: 'Get AMI Set',
+  stability:  base.API.stability.stable,
+  description: [
+    'Retreive a copy of the requested AMI set.',
+  ].join('\n'),
+}, async function (req, res) {
+  let id = req.params.id;
+
+  let amiSet;
+  try {
+    amiSet = await this.AmiSet.load({id: id});
+    res.reply(amiSet.json());
+  } catch (err) {
+    if (err.code === 'ResourceNotFound') {
+      res.status(404).json({
+        error: err.code,
+        msg: id + ' not found',
+      });
+    } else {
+      throw err;
+    }
+  }
+});
+
+api.declare({
+  method: 'post',
+  route: '/ami-set/:id/update',
+  name: 'updateAmiSet',
+  deferAuth: true,
+  scopes: [['aws-provisioner:manage-ami-set:<amiSetId>']],
+  input: 'create-ami-set-request.json#',
+  output: 'get-ami-set-response.json#',
+  title: 'Update AMI Set',
+  stability:  base.API.stability.stable,
+  description: [
+    'Provide a new copy of an AMI Set to replace the existing one.',
+    'This will overwrite the existing AMI Set if there',
+    'is already an AMI Set of that name. This method will return a',
+    '200 response along with a copy of the AMI Set created.',
+    'Note that if you are using the result of a GET on the ami-set',
+    'end point that you will need to delete the lastModified and amiSet',
+    'keys from the object returned, since those fields are not allowed',
+    'the request body for this method.',
+    '',
+    'Otherwise, all input requirements and actions are the same as the',
+    'create method.',
+  ].join('\n'),
+}, async function (req, res) {
+  let input = req.body;
+  let id = req.params.id;
+
+  if (!req.satisfies({id})) {
+    return;
+  }
+
+  let amiSet = await this.AmiSet.load({id: id});
+
+  await amiSet.modify(function(amiS) {
```

Comment:
the name `amiS` sort of implies that it's a plural of amis but with confusing captalization.  Instead, since we're passing in an instance of an AmiSet, let's use the `amiSet` variable name.